src/radiation/moments_source.cpp

```diff
@@ -0,0 +1,801 @@
+#include "radiation/closure.hpp"
+#include "radiation/closure_m1.hpp"
+#include "radiation/closure_mocmc.hpp"
+#include "radiation/local_three_geometry.hpp"
+#include "radiation/radiation.hpp"
+#include "reconstruction.hpp"
+
+//#include "fluid/con2prim_robust.hpp"
+#include "fixup/fixup.hpp"
+#include "fluid/prim2con.hpp"
+
+namespace radiation {
+
+using namespace singularity::neutrinos;
+using fixup::Bounds;
+using singularity::EOS;
+
+template <typename CLOSURE>
+class SourceResidual4 {
+ public:
+  KOKKOS_FUNCTION
+  SourceResidual4(const EOS &eos, const Opacity &opac, const MeanOpacity &mopac,
+                  const Real rho, const Real Ye, const Real bprim[3],
+                  const RadiationType species, /*const*/ Tens2 &conTilPi,
+                  const Real (&gcov)[4][4], const Real (&gammacon)[3][3],
+                  const Real alpha, const Real beta[3], const Real sdetgam,
+                  const Real scattering_fraction, typename CLOSURE::LocalGeometryType &g,
+                  Real (&U_mhd_0)[4], Real (&U_rad_0)[4], const int &k, const int &j,
+                  const int &i)
+      : eos_(eos), opac_(opac), mopac_(mopac), rho_(rho), bprim_(&(bprim[0])),
+        species_(species), conTilPi_(conTilPi), gcov_(&gcov), gammacon_(&gammacon),
+        alpha_(alpha), beta_(&(beta[0])), sdetgam_(sdetgam),
+        scattering_fraction_(scattering_fraction), g_(g), U_mhd_0_(&U_mhd_0),
+        U_rad_0_(&U_rad_0), k_(k), j_(j), i_(i) {
+    lambda_[0] = Ye;
+    lambda_[1] = 0.;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void CalculateMHDConserved(Real P_mhd[4], Real U_mhd[4]) {
+    Real Pg = eos_.PressureFromDensityInternalEnergy(rho_, P_mhd[0] / rho_, lambda_);
+    Real gam1 =
+        eos_.BulkModulusFromDensityInternalEnergy(rho_, P_mhd[0] / rho_, lambda_) / Pg;
+    Real D;
+    Real bcons[3];
+    Real ye_cons;
+    prim2con::p2c(rho_, &(P_mhd[1]), bprim_, P_mhd[0], lambda_[0], Pg, gam1, *gcov_,
+                  *gammacon_, beta_, alpha_, sdetgam_, D, &(U_mhd[1]), bcons, U_mhd[0],
+                  ye_cons);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void CalculateRadConserved(Real U_mhd[4], Real U_rad[4]) {
+    for (int n = 0; n < 4; n++) {
+      U_rad[n] = (*U_rad_0_)[n] - (U_mhd[n] - (*U_mhd_0_)[n]);
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void CalculateRadConservedFromRadPrim(Real P_rad[4], Real U_rad[4]) {
+    PARTHENON_FAIL("Not implemented!");
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  ClosureStatus CalculateRadPrimitive(Real P_mhd[4], Real U_rad[4], Real P_rad[4]) {
+    const Real E = U_rad[0] / sdetgam_;
+    Vec cov_F{U_rad[1] / sdetgam_, U_rad[2] / sdetgam_, U_rad[3] / sdetgam_};
+    Vec cov_H;
+    Real W = 0.;
+    SPACELOOP2(ii, jj) { W += (*gcov_)[ii + 1][jj + 1] * P_mhd[ii + 1] * P_mhd[jj + 1]; }
+    W = std::sqrt(1. + W);
+    Vec con_v{P_mhd[1] / W, P_mhd[2] / W, P_mhd[3] / W};
+    // Use gamma_max from code?
+    if (W > 100) {
+      printf("W = %e! [%i %i %i]\n", W, k_, j_, i_);
```

Comment:
print to stderr or do debug warn or debug fail?

Reply:
I'm just going to remove this for now, if we end up sourcing a very fast fluid we'll just clamp the fluid velocity subsequently (and we aren't really using this 4D solve right now anyway).